File: Python/LinkedListsPractice/linked_list/linked_list.py

class Node:
    def __init__(self, data):
        self.next = None
        self.data = data

    def __str__(self):
        return f"Data: {self.data} linked to ${str(self.next)}"


class LinkedList:

    def __init__(self):
        self.head = None
        self.tail = None
        self.max_size = 100

    def __str__(self):
        return f"Linked list with {self.max_size} nodes"

    def add(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
        else:
            self.tail.next = new_node
        self.tail = new_node

    def contains(self, data):
        if self.head is not None:
            current_node = self.head
            while current_node.next is not None:
                if current_node.data == data:
                    return True
                current_node = current_node.next
        return False

    def get(self, index):
        current_node = self.head
        if current_node is not None and index >= 0:
            current_index = 0
            while current_index < index and current_node is not None:
                if current_node.next is not None:
                    current_node = current_node.next
                    current_index += 1
                else:
                    return None
            return current_node.data
        return None

    def insert_at(self, index, data):
        new_node = Node(data)
        if index ==0:
            new_node.next = self.head
            self.head = new_node
        else:
            previous_node = None
            current_node = self.head
            current_node_index = 0

            while current_node_index < index:
                previous_node = current_node
                current_node = current_node.next
                current_node_index += 1

            new_node.next = current_node
            previous_node.next = new_node

    def print_all(self):
        current_node = self.head
        while current_node is not None:
            #print(str(current_node))
            print(current_node.data)
            current_node = current_node.next

    def remove_from(self, index):
        if index == 0:
            if self.head.next is not None:
                self.head = self.head.next
            else:
                self.head = None
        else:
            previous_node = None
            current_node = self.head
            current_node_index = 0

            while current_node_index < index:
                previous_node = current_node
                current_node = current_node.next
                current_node_index += 1
            previous_node.next = current_node.next

File: Typescript/LinkedLists/linkedList/linked-list.js
"use strict";
class ListNode {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
    getData() {
        return this.data;
    }
    getNext() {
        return this.next;
    }
}
class LinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
    }
    add(data) {
        const newNode = new ListNode(data = data);
        if (this.head == null) {
            this.head = newNode;
        }
        else if (this.tail != null) {
            this.tail.next = newNode;
        }
        this.tail = newNode;
    }
    contains(data) {
        let currentNode = this.head;
        while (currentNode != null) {
            if (currentNode.data === data) {
                return true;
            }
            currentNode = currentNode.next;
        }
        return false;
    }
    get(data) {
        let currentNode = this.head;
        while (currentNode != null) {
            if (currentNode.data === data) {
                return currentNode;
            }
            currentNode = currentNode.next;
        }
        return null;
    }
    getAtIndex(index) {
        let currentNode = this.head;
        let currentIndex = 0;
        if (index == 0) {
            if ((currentNode === null || currentNode === void 0 ? void 0 : currentNode.getData()) != null) {
                return currentNode;
            }
        }
        else {
            while (currentNode != null && currentIndex < index) {
                currentNode = currentNode.getNext();
                currentIndex++;
                if ((currentNode === null || currentNode === void 0 ? void 0 : currentNode.getData()) != null && currentIndex === index) {
                    return currentNode;
                }
            }
        }
        return null;
    }
    insertAt(index, data) {
        console.log(`Calling insert with index ${index} and data ${data}`);
        if (index == 0) {
            const newNode = new ListNode(data);
            newNode.next = this.head;
            this.head = newNode;
        }
        let currentNode = this.head;
        let previousNode = null;
        let currentIndex = 0;
        while (currentIndex < index && currentNode != null) {
            previousNode = currentNode;
            currentNode = currentNode.next;
            currentIndex++;
        }
        console.log(`Current node: ${currentNode}, current index: ${currentIndex}, previous node: ${previousNode}`);
        if (currentIndex == index && previousNode != null) {
            const newNode = new ListNode(data);
            newNode.next = currentNode;
            previousNode.next = newNode;
            if (currentNode == null) {
                this.tail = newNode;
            }
        }
    }
    printAll() {
        let currentNode = this.head;
        while (currentNode != null) {
            console.log(currentNode.data);
            currentNode = currentNode.next;
        }
    }
    removeAt(index) {
        console.log(`Remove method; index: ${index}`);
        let currentNode = this.head;
        if (index === 0) {
            if (currentNode !== null && (currentNode === null || currentNode === void 0 ? void 0 : currentNode.getNext()) !== null) {
                this.head = currentNode.next;
                currentNode = null;
            }
        }
        else {
            let currentNode = this.head;
            let previousNode = null;
            let currentIndex = 0;
            while (currentIndex < index && currentNode != null) {
                previousNode = currentNode;
                currentNode = currentNode.next;
                currentIndex++;
            }
            console.log(`Current node: ${currentNode}, current index: ${currentIndex}, previous node: ${previousNode}`);
            if (currentIndex == index && previousNode != null && currentNode !== null) {
                previousNode.next = currentNode === null || currentNode === void 0 ? void 0 : currentNode.getNext();
                // check whether of not the last node has been removed
                if (previousNode.next == null) {
                    // if so, then reset the tail
                    this.tail = previousNode;
                }
            }
        }
    }
}
module.exports = { ListNode, LinkedList };
